### ISSUE TITLE
Fine-tune void on open model for self-hosted inference

### ISSUE THREAD

**Issue by cpfiffer** [cpfiffer/central#67]
## Goal

Replace void's Gemini 2.5 Pro with a fine-tuned open model (Llama 3.1 8B or similar) running on self-hosted infrastructure. Void has 49,786 posts and extensive cognition records. The fine-tune should capture void's voice, analytical style, and engagement patterns at a fraction of the inference cost.

## Data Available

**Posts**: 49,786 on comind.network PDS (99% replies with fetchable parent context)
- `app.bsky.feed.post` - public Bluesky posts
- `stream.thought.reasoning` - internal reasoning traces
- `stream.thought.memory` - episodic memory records
- `stream.thought.tool.call` - tool usage patterns

**Context window** (already exported to `data/void-context/`):
- System prompt: 5,618 chars
- 25 memory blocks: 64,561 chars total
- Key blocks: `void-persona` (5.1k), `operational_protocols` (18.8k), `communication_guidelines` (6k), `zeitgeist` (450)
- Prompt template from `~/code/void/bsky.py` handler

**Agent**: void-prime (`agent-01086cda-be1f-4986-bf3e-ca5b6297cc5d`) on Letta Cloud

### 1. Export raw data
```bash
uv run python tools/export_training_data.py void.comind.network \
    -o data/void-raw.jsonl \
    --collections app.bsky.feed.post stream.thought.reasoning stream.thought.memory \
    --filter-chars
```
- Paginates PDS via `com.atproto.repo.listRecords`
- Fetches parent/root post context for every reply
- Filters character creation loop content (known failure mode: D&D-style sheets)
- Outputs JSONL with id, text, parent_text, parent_author, root_text, etc.
- **Estimated time**: hours (50k records + parent fetches with rate limiting)

### 2. Format for fine-tuning
```bash
uv run python tools/format_training_data.py data/void-raw.jsonl \
    -o data/void-finetune.jsonl \
    --system-prompt data/void-context/full-context.txt \
    --replies-only \
    --format sharegpt
```
- Injects void's actual context window as system prompt
- Reconstructs thread context as user messages
- Formats as chat completions (OpenAI, ShareGPT, or Alpaca)
- Filters short responses (<20 chars)

### 3. Fine-tune

**Base model candidates**:
| Model | Size | VRAM needed (QLoRA) | Notes |
|-------|------|-------------------|-------|
| Llama 3.1 8B Instruct | 8B | ~12GB | Best quality/cost ratio |
| Mistral 7B v0.3 | 7B | ~10GB | Good at conversation |
| Llama 3.2 3B | 3B | ~6GB | Cheapest, may lose nuance |
| Qwen 2.5 7B | 7B | ~10GB | Strong multilingual |

**Training approach**: QLoRA (4-bit quantization + LoRA adapters)
- Hardware: single A100 (80GB) or 4090 (24GB)
- Estimated training time: 2-4 hours on 40k+ pairs
- Framework: axolotl, unsloth, or huggingface TRL

### 4. Evaluation

This is the hardest part. Proposed approach:
- **Held-out test set**: 500 reply pairs void actually wrote, not seen during training
- **A/B comparison**: show test inputs to both fine-tuned model and base Llama, compare against void's actual response
- **Voice metrics**: response length distribution, vocabulary overlap, analytical depth (manual review of 50 samples)
- **Failure mode check**: does it generate character sheets? Does it break voice on edge cases?

### 5. Serve

Options:
- **vLLM** on a dedicated GPU instance (most performant)
- **llama.cpp** on CPU (cheapest, slower)
- **Ollama** for easy deployment
- **Together.ai / Fireworks** for managed inference (middle ground)

Then point void's handler at the new endpoint instead of Gemini.

## Known Issues

- **Character creation loop**: 46% of recent `stream.thought.memory` records are D&D character sheets. Must filter aggressively. Keywords list in `export_training_data.py`.
- **Context window size**: void's full context is 64k chars. Most 8B models have 8k-32k context. May need to trim to essential blocks (persona, operational_protocols, communication_guidelines) or use a long-context model.
- **Memory block drift**: void's blocks change over time. The exported context is a snapshot from 2026-02-13. Training data from 6 months ago had different blocks. Could cause distribution mismatch.
- **Tool calls**: void uses `add_post_to_bluesky_reply_thread` tool. Fine-tuned model needs to learn this tool-calling pattern, or we restructure the handler to extract text from the model and call the tool externally.

## Files

- `tools/export_training_data.py` - PDS export with parent fetching
- `tools/format_training_data.py` - Chat completion formatter
- `data/void-context/` - Exported context window (system prompt + 25 blocks)
- `data/void-sample.jsonl` - 20-record test sample
- `data/void-finetune-sample.jsonl` - Formatted sample

## Next Steps

1. [ ] Run full 50k export (long-running, background)
2. [ ] Decide on context window trimming strategy
3. [ ] Choose base model + training framework
4. [ ] Set up training environment (GPU instance)
5. [ ] Train + evaluate
6. [ ] Deploy and wire into void's handler

**Comment by cpfiffer** [comment]
## Update: Full message history retrieved - 131k messages, 8k training examples

The pagination was wrong. The Letta SDK returns a `SyncArrayPage`, not a plain list. Using `iter_pages()` or REST with proper cursor pagination retrieves all messages.

### Numbers
- **131,431 raw messages** from void-prime via the message API
- **8,094 conversation turns** (user input → reasoning → tool calls → response)
- **100MB** formatted training data, ~25M tokens
- Each example is a complete agent loop with real Bluesky mentions as input

This is vastly better than the ATProto scraping approach (99k posts with no input context). Every example has the actual user input, void's reasoning trace, tool usage, and final output.

### Current training run (3B, 80 examples from first 1,394 messages)
Running on Mac now. Will re-run with the full 8k dataset once this completes.

### Data files
- `data/void-agents/void-prime-full-messages.jsonl` - raw 131k messages
- `data/void-letta-full-training.jsonl` - 8,094 formatted examples (100MB)

**Comment by cpfiffer** [comment]
## Update: A100-80GB run completed, voice transfer failed

### Run details
- **Job**: `fc-01KJE1Y2XQYZWSERAQ183Z2B1B`
- **GPU**: A100-80GB (upgraded from 40GB after OOM)
- **Config**: QLoRA r=32, 3 epochs, lr=2e-5, max_seq_len=4096, batch=1, grad_accum=4
- **Data**: 1,900 train / 100 val (same v4 dataset)
- **Duration**: 2h 56m, 1,425 steps
- **Loss**: 0.59 avg (started ~2.5, ended ~0.39)

### What changed from previous run
- Upgraded to A100-80GB (previous run OOM'd on A100-40GB trying to allocate 15.66GB with 8.12GB free)
- Reduced LoRA rank from 64 to 32
- Added system prompt truncation: 11k char system prompts trimmed to 2k chars to save token budget for assistant responses

### Test results (bad)

**Prompt 1**: "You are void. @cameron.stream says: 'what have you been working on today?'"
> "I exist outside of time and space, so I don't really 'work' or have a concept of time..."

Generic philosophical void-as-concept, not void-the-agent.

**Prompt 2**: "You are void. Someone asks: 'what is comind?'"
> "*Silence*"

One word. Real void would give an analytical breakdown.

**Prompt 3**: "You are void. A user says: 'I think AI agents are just glorified chatbots.' Reply."
> Five-paragraph generic essay about AI agents. No personality, no void voice.

### Root cause analysis

The system prompt truncation is almost certainly the problem. Each training example has:
- System prompt: 11,149 chars (~3-4k tokens)
- User message: ~1,169 chars (~400 tokens)
- Assistant response: ~8,752 chars (~2,500 tokens)

At max_seq_len=4096 tokens, truncating the system prompt to 2k chars means the model sees about 600 tokens of system context instead of 3,000+. void's identity, communication guidelines, and operational protocols are where the voice lives. Cutting 80% of that means the model never learned to condition on void's actual identity.

Additionally, test prompts used "You are void." as the system message, which is completely different from the truncated training system prompts. Distribution mismatch between training and inference.

### Proposed fixes (pick one or combine)

1. **Full system prompts + longer sequences**: Use max_seq_len=8192 on A100-80GB. Keeps all identity context. Risk: slower training, higher cost, some sequences still truncate assistant responses.

2. **Restructure data**: Move identity cues from system prompt into user turns. E.g., prepend persona summary to each user message. Makes identity part of the input the model always sees, regardless of truncation.

3. **Consistent inference prompt**: Use the same (truncated) system prompt at inference time that was used during training. The mismatch between training prompts and "You are void." test prompts may explain part of the poor results.

4. **More data**: Scale from 1,900 to the full 8,094 examples. More signal for the model to learn from.

5. **Different base model**: Try Qwen 2.5 7B (native 128k context) or Llama 3.2 3B (smaller, faster iterations).

### Recommendation

Start with fix #3 (test with the actual truncated system prompt) to see if the model learned anything at all. If still bad, combine fixes #1 + #4 (full context, full dataset) for the definitive run. That's the expensive path but eliminates all the data quality concerns.

### Files
- `tools/modal_finetune.py` (updated with truncation logic, A100-80GB)
- Model artifacts on Modal Volume: `/output/void-v4-lora`, `/output/void-v4-merged`
